src/main/java/io/temporal/internal/statemachines/CancelExternalStateMachine.java

```diff
@@ -0,0 +1,127 @@
+      Functions.Proc1<NewCommand> commandSink) {
+    new CancelExternalStateMachine(attributes, completionCallback, commandSink);
+  }
+
+  private CancelExternalStateMachine(
+      RequestCancelExternalWorkflowExecutionCommandAttributes requestCancelAttributes,
+      Functions.Proc2<Void, RuntimeException> completionCallback,
+      Functions.Proc1<NewCommand> commandSink) {
+    super(newStateMachine(), commandSink);
+    this.requestCancelAttributes = requestCancelAttributes;
+    this.completionCallback = completionCallback;
+    action(Action.SCHEDULE);
+  }
+
+  enum Action {
+    SCHEDULE
+  }
+
+  enum State {
+    CREATED,
+    REQUEST_CANCEL_EXTERNAL_COMMAND_CREATED,
+    REQUEST_CANCEL_EXTERNAL_COMMAND_RECORDED,
+    CANCEL_REQUESTED,
+    REQUEST_CANCEL_FAILED,
+  }
+
+  private static StateMachine<State, Action, CancelExternalStateMachine> newStateMachine() {
+    return StateMachine.<State, Action, CancelExternalStateMachine>newInstance(
+            "CancelExternal", State.CREATED, State.CANCEL_REQUESTED, State.REQUEST_CANCEL_FAILED)
+        .add(
+            State.CREATED,
+            Action.SCHEDULE,
+            State.REQUEST_CANCEL_EXTERNAL_COMMAND_CREATED,
+            CancelExternalStateMachine::createCancelExternalCommand)
+        .add(
+            State.REQUEST_CANCEL_EXTERNAL_COMMAND_CREATED,
+            CommandType.COMMAND_TYPE_REQUEST_CANCEL_EXTERNAL_WORKFLOW_EXECUTION,
+            State.REQUEST_CANCEL_EXTERNAL_COMMAND_CREATED)
+        .add(
+            State.REQUEST_CANCEL_EXTERNAL_COMMAND_CREATED,
+            EventType.EVENT_TYPE_REQUEST_CANCEL_EXTERNAL_WORKFLOW_EXECUTION_INITIATED,
+            State.REQUEST_CANCEL_EXTERNAL_COMMAND_RECORDED)
+        .add(
+            State.REQUEST_CANCEL_EXTERNAL_COMMAND_RECORDED,
+            EventType.EVENT_TYPE_EXTERNAL_WORKFLOW_EXECUTION_CANCEL_REQUESTED,
+            State.CANCEL_REQUESTED,
+            CancelExternalStateMachine::notifyCompleted)
+        .add(
+            State.REQUEST_CANCEL_EXTERNAL_COMMAND_RECORDED,
+            EventType.EVENT_TYPE_REQUEST_CANCEL_EXTERNAL_WORKFLOW_EXECUTION_FAILED,
+            State.REQUEST_CANCEL_FAILED,
+            CancelExternalStateMachine::notifyFailed);
+  }
+
+  private void createCancelExternalCommand() {
+    addCommand(
+        Command.newBuilder()
+            .setCommandType(CommandType.COMMAND_TYPE_REQUEST_CANCEL_EXTERNAL_WORKFLOW_EXECUTION)
+            .setRequestCancelExternalWorkflowExecutionCommandAttributes(requestCancelAttributes)
+            .build());
+  }
+
+  private void notifyCompleted() {
+    completionCallback.apply(null, null);
+  }
+
+  private void notifyFailed() {
+    WorkflowExecution execution =
+        WorkflowExecution.newBuilder()
+            .setWorkflowId(requestCancelAttributes.getWorkflowId())
+            .setRunId(requestCancelAttributes.getRunId())
+            .build();
+    completionCallback.apply(null, new CancelExternalWorkflowException(execution, "", null));
```

Comment:
we don't have any information as to why cancellation failed?

Reply:
Added message.